### DISCUSSION TITLE
How to work out which customization are not used

### DISCUSSION THREAD

**Discussion by ricardo-silva-titian** [AutoFixture/AutoFixture#1391] [Q&A]
Suppose you have the following scenario:
- complex chain of dependencies  in a namespace where many classes share the same interface
- significant number of customization
- the Fixture is working. It is producing the class instances that you need

However you made many changes and you are not sure if all customization are still needed?
Is there a way to know (programmatically) which customizations were not used by the fixture?

**Comment by aivascu** [comment]
@ricardo-silva-titian can't think at the moment of any way to "tree shake" the customization graph.
I'm not sure it is even possible since customizations might be adding things to any part of the fixture graph.

I would approach this as if I was trying to find dead production code. That might include static analysis tools, tracing, logs, etc.
Also I would probably refactor the customizations into small reusable parts that can be easily composed and reasoned about.
While doing so mabybe apply refactoring patterns like the mikado method, to ensure it is possible to revert to a previous state in case things break.

In the end AutoFixture is a very open-ended tool, and ones experience with it is as good as the design of the test code.

Perhaps if you would post some examples of how you use AutoFixture, I might be able to give you more specific advice.
